moquette-0.17/broker/src/main/java/io/moquette/broker/unsafequeues/QueuePool.java

```diff
@@ -0,0 +1,490 @@
+
+        private SegmentAllocationCallback(QueuePool queuePool) {
+            this.queuePool = queuePool;
+        }
+
+        @Override
+        public void segmentedCreated(String name, Segment segment) {
+            queuePool.segmentedCreated(name, segment);
+        }
+    }
+
+    private void segmentedCreated(String name, Segment segment) {
+        LOG.debug("Registering new segment {} for queue {}", segment, name);
+        final QueueName queueName = new QueueName(name);
+        List<SegmentRef> segmentRefs = this.queueSegments.computeIfAbsent(queueName, k -> new LinkedList<>());
+
+        // adds in head
+        segmentRefs.add(0, new SegmentRef(segment));
+
+        LOG.debug("queueSegments for queue {} after insertion {}", queueName, segmentRefs);
+    }
+
+    public static QueuePool loadQueues(Path dataPath, int pageSize, int segmentSize) throws QueueException {
+        // read in checkpoint.properties
+        final Properties checkpointProps = createOrLoadCheckpointFile(dataPath);
+
+        // load last references to segment and instantiate the allocator
+        final int lastPage = Integer.parseInt(checkpointProps.getProperty("segments.last_page", "0"));
+        final int lastSegment = Integer.parseInt(checkpointProps.getProperty("segments.last_segment", "0"));
+
+        final PagedFilesAllocator allocator = new PagedFilesAllocator(dataPath, pageSize, segmentSize, lastPage, lastSegment);
+
+        final QueuePool queuePool = new QueuePool(allocator, dataPath, segmentSize);
+        queuePool.loadQueueDefinitions(checkpointProps);
+        LOG.debug("Loaded queues definitions: {}", queuePool.queueSegments);
+
+        queuePool.loadRecycledSegments(checkpointProps);
+        LOG.debug("Recyclable segments are: {}", queuePool.recycledSegments);
+        return queuePool;
+    }
+
+    public Set<String> queueNames() {
+        return queues.keySet().stream().map(qn -> qn.name).collect(Collectors.toSet());
+    }
+
+    private static Properties createOrLoadCheckpointFile(Path dataPath) throws QueueException {
+        final Path checkpointPath = dataPath.resolve("checkpoint.properties");
+        if (!Files.exists(checkpointPath)) {
+            LOG.info("Can't find any file named 'checkpoint.properties' in path: {}, creating new one", dataPath);
+            final boolean notExisted;
+            try {
+                notExisted = checkpointPath.toFile().createNewFile();
+            } catch (IOException e) {
+                LOG.error("IO Error creating the file {}", checkpointPath, e);
+                throw new QueueException("Reached an IO error during the bootstrapping of empty 'checkpoint.properties'", e);
+            }
+            if (!notExisted) {
+                LOG.warn("Found a checkpoint file while bootstrapping {}", checkpointPath);
+            }
+        }
+
+        final FileReader fileReader;
+        try {
+            fileReader = new FileReader(checkpointPath.toFile());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fileReader*. Execution paths that do not contain closure statements were detected. Ensure *fileReader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).